File: 05-binary_search/00033-Search_in_Rotated_Sorted_Array/00033-search_in_rotated_sorted_array.java
// 33. Search in Rotated Sorted Array
// https://leetcode.com/problems/search-in-rotated-sorted-array/


class Solution {
    // SOLUTION
    public int search(int[] nums, int target) {
        int l = 0;
        int h = nums.length;

        while (l < h) {
            int m = (l+h) / 2;
            if (target < nums[0] && nums[0] < nums[m])
                l = m+1;
            else if (target >= nums[0] && nums[0] > nums[m])
                h = m;
            else if (nums[m] < target)
                l = m+1;
            else if (nums[m] > target)
                h = m;
            else
                return m;
        }

        return -1;
    }


    public static void main(String[] args) {
        Solution o = new Solution();
        
        // INPUT
        int[] nums = {4,5,6,7,0,1,2};
        int target = 0;

        // OUTPUT
        var result = o.search(nums, target);
        System.out.println(result);
    }
}

File: 05-binary_search/00033-Search_in_Rotated_Sorted_Array/00033-search_in_rotated_sorted_array.go
// 33. Search in Rotated Sorted Array
// https://leetcode.com/problems/search-in-rotated-sorted-array/

package main

import "fmt"

// SOLUTION
func search(nums []int, target int) int {
    l, h := 0, len(nums)

    for (l < h) {
        m := (l+h) / 2
        if (target < nums[0] && nums[0] < nums[m]) {
            l = m+1
        } else if (target >= nums[0] && nums[0] > nums[m]) {
            h = m
        } else if (nums[m] < target) {
            l = m+1
        } else if (nums[m] > target) {
            h = m
        } else {
            return m
        }
    }

    return -1
}


func main() {
    // INPUT
    nums := []int{4,5,6,7,0,1,2}
    target := 0
    
    // OUTPUT
    result := search(nums, target)
    fmt.Println(result)
}